src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -3407,317 +3557,470 @@ private static string GetTimeZoneDirectory()
-                                                    GmtTime[dstStartTypeIndex],
-                                                    out startDate);
+                // n
+                // This specifies the Julian day, with n between 0 and 365.February 29 is counted in leap years. 
 
-            //
-            // Choosing the endDate based on the startDate:
-            //
-            // startTransitionDate.Month -> end
-            // 1        4|5        8|9       12
-            // [-> 06/15]|[-> 11/15]|[-> 06/15]
-            //
-            int startDateMonth = startDate.Month;
-            int startDateYear  = startDate.Year;
+                // These two rules cannot be expressed with the current AdjustmentRules
+                // One of them *could* be supported if we relaxed the TransitionTime validation rules, and allowed
+                // "IsFixedDateRule = true, Month = 0, Day = n" to mean the nth day of the year, picking one of the rules above
 
-            if (startDateMonth <= 4) {
-                endDate = new DateTime(startDateYear,   06, 15);
-                dstEnd = c_transition7_15;
-            } else if (startDateMonth <= 8) {
-                endDate = new DateTime(startDateYear,   11, 15);
-                dstEnd = c_transition12_15;
-            }
-            else if (startDateYear < 9999) {
-                endDate = new DateTime(startDateYear+1, 06, 15);
-                dstEnd = c_transition7_15;
-            }
-            else {
-                endDate = DateTime.MaxValue;
-                dstEnd = c_transition7_15;
+                throw new InvalidTimeZoneException(Environment.GetResourceString("InvalidTimeZone_JulianDayNotSupported"));
             }
+        }
 
-            AdjustmentRule r = AdjustmentRule.CreateAdjustmentRule(startDate, endDate, daylightBias, dstStart, dstEnd);
-            rulesList.Add(r);
-        }         
+        /// <summary>
+        /// Parses a string like Mm.w.d into month, week and DayOfWeek values.
+        /// </summary>
+        /// <returns>
+        /// true if the parsing succeeded; otherwise, false.
+        /// </returns>
+        private static bool TZif_ParseMDateRule(string dateRule, out int month, out int week, out DayOfWeek dayOfWeek)
+        {
+            month = 0;
+            week = 0;
+            dayOfWeek = default(DayOfWeek);
 
+            if (dateRule[0] == 'M')
+            {
+                int firstDotIndex = dateRule.IndexOf('.');
+                if (firstDotIndex > 0)
+                {
+                    int secondDotIndex = dateRule.IndexOf('.', firstDotIndex + 1);
+                    if (secondDotIndex > 0)
+                    {
+                        string monthString = dateRule.Substring(1, firstDotIndex - 1);
+                        string weekString = dateRule.Substring(firstDotIndex + 1, secondDotIndex - firstDotIndex - 1);
+                        string dayString = dateRule.Substring(secondDotIndex + 1);
```

Comment:
I assume it'll be uncommon for monthString/weekString/etc. not to parse correctly?  If it is common, though, consider moving the weekString and dayString creation into the if blocks just before they're needed, so we don't create those strings unnecessarily.